sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -498,14 +511,24 @@ static void _process_coap_pdu(gcoap_socket_t *sock, sock_udp_ep_t *remote, sock_
                         _cache_process(memo, &pdu);
                     }
                 }
+
+                bool observe_notification = coap_has_observe(&pdu);
+
                 if (memo->resp_handler) {
                     memo->resp_handler(memo, &pdu, remote);
                 }
 
                 if (memo->send_limit >= 0) {        /* if confirmable */
                     *memo->msg.data.pdu_buf = 0;    /* clear resend PDU buffer */
                 }
-                memo->state = GCOAP_MEMO_UNUSED;
+
+                /* The memo must be kept if the response is an observe notification.
+                 * Non-2.xx notifications indicate that the associated observe entry
+                 * was removed on the server side. Then also remove it here. */
+                if (!observe_notification || (code_class != COAP_CLASS_SUCCESS)) {
+                    memo->state = GCOAP_MEMO_UNUSED;
```

Comment:
I fail to find the binding between the comment `.. remove it here.` and the code `foo = MEMO_UNUSED`. Could you clarify?

Reply:
`foo = GCOAP_MEMO_UNUSED` "removes" the memo.  I tried to clarify with better comments and replacing "remove" with "free".